### DISCUSSION TITLE
How to load a program on a Franka Emika Panda?

### DISCUSSION THREAD

**Discussion by K4rl33** [visose/Robots#253] [Q&A] [answered]
Hello! 
I created a program for the panda with robots and I would like to load and execute it on the robot. I only have the "desk" web interface. Can I do it? If not, what do I need to load it and play?
Thanks a lot,
BB!

**Comment by visose** [comment, marked as answer]
Hi, currently it's a bit complicated to set it up, this is a quick summary:
- You need a Linux machine with the real-time kernel enabled (it's been tested with Ubuntu 22.04).
- The linux machine should also have 2 ethernet adapters, one connects directly to the FCI port of the robot and one to a Windows PC or Mac that has Grasshopper + Robots (this one could also be a wifi/router/etc).
- In the Linux machine, you should install this: https://github.com/visose/frankx (note that this is a fork of the original Frankx project, you must install the fork not the original).
- The Linux machine should have SSH and sFTP enabled.

You can then use it in the following way:
- Enable FCI using the desk web interface.
- The Create Program component will generate Python code that works with the Frankx library.
- You can then use the Remote component to upload this python script to the Linux machine and to start the program. It uses sFTP to upload it and SSH to run it. The IP input of the Remote component should have this format: `sftp://user:password@192.168.48.210/home/user/folder`, you should replace user, password, the IP, and folder.

**Comment by UnfoldAntwerp** [comment]
Hi @visose, 

Pretty excited to finally try out the support for Franka Panda after being tipped by @KonradJuenger last year. Few questions:
-Looking at the instructions and the hardware requirements for Ubuntu 22.04 with RT kernel I noticed that it will run on a Raspberry Pi 4 or 5. Any idea if this would be sufficient? Would greatly simplify the setup.
-The wording above isn't entirely clear. Do we need two network connections, one wired ethernet to Panda + one more (wired or wireless) to upload python script programs. Or do we really need two ethernet wired connections?

thanks!

**Reply by KonradJuenger** [reply]
jumping in here as I recently worked with FE panda and robots, Running the control loop on a raspberry pi might be possible but could be more hassle than it's worth, e.g. compiling the rt preempt kernel extension on a slow cpu, introducing a lot of unknowns (arm/x86 compatibility). in my experience it is hassle enough getting it running as intended 

for networking there are 2 connections necessary: one wired directly to the FCI box and one ( wired or wireless) to the Windows PC running Rhino. Franka Emika told me 2019 they advice for some specific "server grade" intel network card, which I bought at the time -- but I think I should be running fine with a standard card. When there is too much latency (over 1ms for some time) the robot will go into e-stop or for any other reason haha. (not trying to discourage anyone, but every time I tried something specific with the FE panda I lost many hairs, as the documentation and accessibility is so poor)

**Reply by visose** [reply]
Hi, I agree with @KonradJuenger, we never tested it with a Raspberry Pi. What I would recommend is a small PC box with an x86-64 CPU that is compact enough to fit in your robot cell, no need to be high performance. Not as cheap as a Raspberry Pi, but are still pretty cheap.
It should have a dedicated network card for the real time communication with the robot.
To connect with your Windows or Mac machine running Rhino, the best if it also has a wifi adapter and connect it to a router that the Rhino computer can also connect to.

**Reply by UnfoldAntwerp** [reply]
Thanks for the input! @KonradJuenger I wouldn't think about this as a viable option weren't it for the Ubuntu Pro page to specifically mention running it on Intel or RaspPi: https://ubuntu.com/real-time. So in my mind that means it is a fairly hassle free packaged solution no? But I agree that the safer route is the on of lesser resistance and possible roadblocks. @visose a mini pc could nicely integrate too. I'll keep you posted.

**Reply by visose** [reply]
The main problem is that this integration relies on a fork of the Frankx project that we have only compiled to x86-64. You would have to compile it to ARM, which will likely involve compiling some of its dependencies (other libraries this project relies on) to ARM as well, which might also have other dependencies, and so on recursively. In some cases you might have to make manual changes to get them to compile. Hopefully all dependencies are either open source or provide a compiled ARM binary.
